### PR TITLE
552 Correct ParameterSet constructor

### PR DESCRIPTION
**Pull request by mknaranja**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Make sure that the [pre-commit linting/style checks pass](https://github.com/DLR-SC/memilio/wiki).

If functions were changed or functionality was added:
* [x] Tests for new functionality has been added
* [x] A local test was succesful

If new functionality was added:
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [ ] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [ ] Are new methods referenced? Did you provide further documentation? Has the glossary been updated? 

The following questions are addressed in the documentation if need be: 
* [x] Developers (what did you do?, how can it be maintained?)
* [x] For users (how to use your work?)
* [ ] For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [x] Coverage report for new code is acceptable.

### CODE REVIEW

**Comment by DavidKerkmann** [comment]
I was able to follow your explanation, but I don't understand why this constructor is chosen since it should be ruled out due to the absence of the get_default function, as you pointed out. Can we discuss this tomorrow?
If I can follow this, then I understand that this solves the issue, but it would not solve it if you pass anything else, would it?

**Comment by mknaranja** [comment]
I just added some default tests (which dont do a lot) and changed some names

**Comment by mknaranja** [comment]
@DavidKerkmann @jubicker @HenrZu @xsaschako Could this be validated by at least two of you?

**Review by DavidKerkmann — COMMENTED** [review]
Thank you, I only have one more question in the comment!

File: cpp/tests/test_odeseir.cpp [hunk 1]
```diff
@@ -25,6 +25,18 @@
 #include "memilio/compartments/simulation.h"
 #include <gtest/gtest.h>
 
+TEST(TestSeir, simulateDefault)
```
Comment (jubicker) [inline cpp/tests/test_odeseir.cpp:28]: Shouldn't it be TestOdeSeir?

Comment (mknaranja) [reply cpp/tests/test_odeseir.cpp:28]: Yes. but everywhere in the file. Should be changed with some future commit

File: cpp/tests/test_ide_seir.cpp [hunk 2]
```diff
@@ -25,7 +25,29 @@
...
+        init.add_time_point(init.get_last_time() + dt, Vec::Constant(1, 10.));
+    }
+
+    mio::iseir::IdeSeirModel model(std::move(init), dt, 10);
+    model.simulate(tmax);
+    auto result = model.calculate_EIR();
+
+    // TODO: It is actually an inconsistency that the simulation of the IDE model goes to tmax + dt,
+    // instead of tmax; this has to be corrected in the future.
+    EXPECT_NEAR(result.get_last_time(), (double)tmax + 0.1, 1e-10);
```
Comment (xsaschako) [inline cpp/tests/test_ide_seir.cpp:47]: Please change 0.1 to dt.

File: cpp/tests/test_ide_seir.cpp [hunk 3]
```diff
@@ -25,7 +25,29 @@
...
+    init.add_time_point<Eigen::VectorXd>(-12.0, Vec::Constant(1, 10.));
+    while (init.get_last_time() < 0) {
+        init.add_time_point(init.get_last_time() + dt, Vec::Constant(1, 10.));
+    }
+
+    mio::iseir::IdeSeirModel model(std::move(init), dt, 10);
+    model.simulate(tmax);
+    auto result = model.calculate_EIR();
+
+    // TODO: It is actually an inconsistency that the simulation of the IDE model goes to tmax + dt,
```
Comment (xsaschako) [inline cpp/tests/test_ide_seir.cpp:45]: Please open an issue

File: cpp/memilio/utils/parameter_set.h [hunk 4]
```diff
@@ -232,11 +236,16 @@ class ParameterSet
...
      * Arguments get forwarded to get_default of parameters.
+     @tparam T1 First argument of get_default(...) function called on the parameters, e.g., T1=AgeGroup&.
+     @tparam TN Further arguments passed to get_default(...) functions. 
      */
     template <class T1, class... TN,
               class = std::enable_if_t<
-                  details::AllOf<details::BindTail<has_get_default_member_function, T1, TN...>::template type,
-                                 ParameterTagTraits<Tags>...>::value>>
+              // Avoid erroneous template deduction for T1=ParameterSet as this constructor could falsely be considered
+              // as a copy constructor for non-const lvalue references.
```
Comment (xsaschako) [inline cpp/memilio/utils/parameter_set.h:245]: Should this comment be here? It's in the middle of code.

**Review by xsaschako — COMMENTED** [review]
Small things to consider. Otherwise looks good.

**Review by DavidKerkmann — APPROVED** [review]
Everything is resolved, ready to merge.

**Pull request merged**